### ISSUE TITLE
stop using finalizers for resource management?

### ISSUE THREAD

**Issue by JeffBezanson** [JuliaLang/julia#11207]
Finalizers are inefficient and unpredictable. And with the new GC, it might take much longer to get around to freeing an object, therefore tying up its resources longer. Ideally releasing external resources should not be tied to how memory management works.

We are already not far from this with the `open(f) do` construct. I think that and/or `with` should be used. Perhaps there could be some other mechanism for registering files to close eventually.

Discussed this with @carnaval .

**Comment by ScottPJones** [comment]
Hmmm... I was just going to start using finalizers (but still had some questions about them to investigate).
My needs are simple: 1 pointer in Julia to a structure allocated / controlled by C, which also contains a pointer possibly allocated by my C code, or else allocated by DBMS allocator.  If the object goes out of scope in Julia, and is about to be GCed, I thought the finalizer would allow me to call my C release code... It concerned me though that finalizers are apparently associated with the object, not the type.

**Comment by quinnj** [comment]
Go has the `defer` keyword, the usage is:

f := os.Open(file)
defer f.Close()

which pushes f.Close() onto a stack of function calls that get evaluated
when the enclosing scope ends.

http://blog.golang.org/defer-panic-and-recover

-Jacob

**Comment by timholy** [comment]
If one is rethinking this, the [machinations](https://github.com/JuliaGPU/CUDArt.jl/blob/107ae758fe5490ebf41569f4e58ec062573adef5/src/pointer.jl#L26-L51) of [CUDArt](https://github.com/JuliaGPU/CUDArt.jl/blob/107ae758fe5490ebf41569f4e58ec062573adef5/src/pointer.jl#L26-L51) to manage GPU memory in a GC-compatible way are probably [amusing fodder for thought](https://github.com/JuliaGPU/CUDArt.jl/blob/107ae758fe5490ebf41569f4e58ec062573adef5/test/gc.jl#L43-L62). The arrival of `finalize` was a huge step forward.

**Comment by carlobaldassi** [comment]
For reference, I'll also add the case of GLPK, which bears similarities with CUDArt, see e.g. [here](https://github.com/JuliaOpt/GLPK.jl/blob/f7c0ecfb16c03e755c2c496c13b504e7c2ddc3ec/src/GLPK.jl#L327-L355) and [here](https://github.com/JuliaOpt/GLPK.jl/blob/f7c0ecfb16c03e755c2c496c13b504e7c2ddc3ec/src/GLPK_checks.jl#L19-L33).

**Comment by JeffBezanson** [comment]
See also #1037

It would be great to get rid of finalizers entirely, but that's probably not realistic. For starters, I would still allow finalizers but not use them to close files and such.

@ScottPJones you can definitely use finalizers to call your C release code.

Finalizers can be associated with a type by adding them to all instances in the constructor :)
Seriously though, I'm not sure how it would work to associate finalizers with types. For example, you can't iterate over all dead objects to see which ones might have finalizers. And how is the right type identified? That's usually done through method calls, but who calls what function, and when, to determine what to finalize? The simplest thing is to just give the GC a list of objects with finalizers attached.

**Comment by elextr** [comment]
@JeffBezanson it is very useful to have a mechanism to allow freeing of limited resources (like file descriptors) as soon as reasonably possible.  As you say finalizers will *eventually* get around to it, but that doesn't prevent exhaustion in the meantime.

One question, are finalizers always run, no matter how the program exits, so its always possible to be sure any resource does not remain locked?

**Comment by ScottPJones** [comment]
@elextr yes - that sort of exhaustion has been a big issue with the sort of code that I write, where it has to stay running with minimal downtime for years...

**Comment by elextr** [comment]
@ScottPJones then its probably best if you do your resource management explicitly yourself, certainly don't rely on anything in the semantics of *any* language, unless specified and guaranteed.

Specifically the semantics of the Julia GC and hence finalizers is not guaranteed, it currently happens to have recently changed to a generational GC in 0.4, but is not generational in 0.3, and that may change in 0.4/0.5 again when threading lands (for example).  All you can know about a finalizer is that the object it relates to is no longer in use when the finalizer is run, but my reading of [this](http://docs.julialang.org/en/latest/stdlib/base/?highlight=finalizer#Base.finalizer) suggests that it may not be run for bitstypes, hence my question above.

**Comment by aviks** [comment]
Another use case is the interaction of the Java and Julia GC's in JavaCall. Objects retrieved from Java into Julia need to explicitly de-referenced in Java when they are no longer used within Julia. This is achieved via the finalizers. Which works fine, except that the Java VM can have greater memory pressure than the Julia VM. In that case, the JVM can run out of memory, before Julia decides that the GC needs to be run.

**Comment by timholy** [comment]
@ScottPJones, I hear you. In several places like HDF5 and CUDArt, the key was to write code like
```
open(filename) do file
    # do stuff
end
```
which guarantees that `file` will be closed (immediately upon completion) even if `stuff` has an error in it. That construct currently has some performance overhead (anonymous functions), but in most cases is worth it. You can manually use `try...finally` in cases where you can't tolerate the overhead.

**Comment by ScottPJones** [comment]
@elextr I should have been clearer... I'm not planning on *relying* on the finalizers at all.
My APIs are identical (with some name changes, can't have ! in C function names), in C11, C++11,
Python, Java, and Julia...
I just want to prevent memory leaks, esp. when people are playing around / prototyping stuff in the REPL.
For example, something like the following:
```julia
myObj = DA.PackedData(1000) # creates a packed data buffer with initial size at least 1000 bytes.
push!(myObj, "Encode a string")
push!(myObj, 5.2332) # encode an IEEE binary floating point number
save!(myDBMS, myObj) # write packed record out as a row
release(myObj) # Release the underlying C buffer object, 0 out the pointer in the Julia myObj object...
```
What happened many times, in the REPL, is I accidentally set myObj to something else before calling release... so I lost memory each time...
Using the finalizer is just to catch stupid things like that...

@timholy That's good to know, but is that sort of syntax only for files?  (sorry, my newbieness with Julia is showing again!)

**Comment by timholy** [comment]
@ScottPJones, it's a standard julia convention, see http://docs.julialang.org/en/latest/manual/functions/#do-block-syntax-for-function-arguments. You have to write a version of your function that takes another function as the first argument (see, e.g., the methods defined for `open`). Internally, it's just `try...finally`.

**Comment by tknopp** [comment]
@ScottPJones No the do syntax is not restricted to files see http://julia.readthedocs.org/en/latest/manual/functions/#do-block-syntax-for-function-arguments

I think this is the standard way to do it in Julia and as @timholy said it is used in various places in Julia land. In Gtk.jl we have also some places.

Where the finalizers are important is when the type goes out of scope. We have for instance in Gtk.jl the situation where it is really needed.

**Comment by tknopp** [comment]
oh Tim is faster, sorry.

**Comment by ScottPJones** [comment]
Thanks @tknopp & @timholy!  Sorry for the noise, I really am trying to memorize the manual, but Julia is such a *large* language!

**Comment by timholy** [comment]
It definitely takes a while, no apologies needed.

**Comment by tknopp** [comment]
@JeffBezanson: What is the actual proposal of this issue? Isn't the do syntax already consistently been used for files? I think the finalizers are useful when the scope is not local.

**Comment by carnaval** [comment]
We probably can't remove finalizers alltogether because then we would be leaking resources. I think this issue is more about conventions on "good practice for resource management" since the biggest problem (besides performance) is that the gc is very lazy : it only works under pressure, that is _memory_ pressure. It has no way to know e.g. how many file descriptor are open by the program, so if your handle object is small, the gc will be completely fine keeping it around for a long time while you exhaust your open fd limit.

I don't have any good idea about this by the way...

**Comment by wildart** [comment]
I found finalizers unreliable. When interfacing with C code, I would really prefer something like Go `defer` rather then use finalizer to release resources. I opt to a manual resource management event though it increase several times amount of code to be written.

**Comment by JeffBezanson** [comment]
@tknopp good question. My proposal would be

- Have a standard "release resources" function, maybe `close`, maybe `finalize`, but the same for every type with this issue.
- Use `with` instead of type-specific functions like `open` for this.
- Document that this should be used instead of relying on finalizers, and use it ourselves everywhere we can.
- Make a concerted effort to remove use of finalizers, e.g. for BigInt
- Don't add finalizers to I/O-related objects by default. Instead maybe the REPL could add finalizers for interactive objects, to call the standard `close` function.

The last item sounds drastic, but as it is finalizers might not be invoked for a *very* long time, and unpredictably. You could still use finalizers as an escape hatch. If you're not sure how to handle releasing some object, you can just call `finalizer(x)` on it any time.

**Comment by tknopp** [comment]
Ok. Is there some issue what `with` is and where it differs from the do syntax?

**Comment by carlobaldassi** [comment]
I'll just add another small issue about using finalizers with IO objects which I very recently discovered: on Windows, trying to call `rm` on a file with an open descriptor fails. This made the [FastaIO tests](https://github.com/carlobaldassi/FastaIO.jl/blob/3f83d892ed537d79d7c309c5494d12e025e926f0/test/runtests.jl#L147-L151) fail, because I was [relying on `finalize` to close the file](https://github.com/carlobaldassi/FastaIO.jl/blob/7a4d7337865633e9905859aaba23fb6d81273a70/src/FastaIO.jl#L219) after I finished reading it, and I was deleting it after the tests. I never noticed the bug since on Linux that works fine.
So this is probably not a very common situation, but — in association with the unpredictability of the GC — may lead to OS-specific, non-deterministic bugs.

**Comment by elextr** [comment]
@JeffBezanson how do you propose to handle objects whose lifetime exceeds the scope of the `with`, eg ones returned from the function?

**Comment by JeffBezanson** [comment]
If an object lifetime exceeds the local scope, you can't use `with`. The only options I see in that case are (1) somebody downstream uses `with`, (2) you add a finalizer to the object before returning it.

**Comment by StefanKarpinski** [comment]
Another idea is to have some types opt into reference counting and finalize them when their counts get to zero. It's not entirely clear to me how to make a mix of refcounting and not work, however.

**Comment by jakebolewski** [comment]
watch out, may be flayed by mentioning reference counting :-)

**Comment by carnaval** [comment]
the problem with mixed refcount is that a refcounted object can still be kept alive by a non refcounted one (worst case : the object keeping it alive is in oldgen). Then you don't get the "immediate finalization" property.

**Comment by carnaval** [comment]
To alleviate the late finalization problem we could also teach the gc about other kind of resources so that it can be taken into account in the collection heuristics. So e.g. you could register a "file descriptor", or "GPU memory" something, and then explicitely say : I allocated X of this, running _this_ finalizer will get me Y of this back.

Painful to implement though. And it can only make gc overhead worse (by collecting more often).

**Comment by StefanKarpinski** [comment]
> the problem with mixed refcount is that a refcounted object can still be kept alive by a non refcounted one (worst case : the object keeping it alive is in oldgen). Then you don't get the "immediate finalization" property.

Yes, in such a scheme every reference that could transitively reach anything refcounted would need to maintain a refcount. That includes most abstract slots, and slots in data structures that can refer to refcounted objects. But that still excludes most things we care about the performance of.

**Comment by ScottPJones** [comment]
Well, mix refcount in some sense is what I'm already doing:  in Julia, I have a type that contains a Ptr{}
to a C allocated (refcounted) structure.  Normally, I call release! on the Julia object, and it calls C to decrement the refcount, and I clear my pointer in the Julia object.  I've been adding finalize calls so that if the Julia object is garbage collected, the refcount is still correct, and the memory gets freed on the C side... (mainly for when people are using the REPL, and might accidently overwrite the variable containing the Julia object).  [I hope this technique seems OK to all of you!]
Note, why can't Julia, when it is freeing an object, simply check to see if there is a finalize! method for that type, and call that?)

**Comment by carnaval** [comment]
@StefanKarpinski which is practically every possible object (modules, generic functions, ...) except some leaf datatypes. I'm not eager to retrofit refcount inside the C codebase (and the generated code) everywhere. Especially since naive refcount is inefficient so we would need to implement deferred refcount on top of that.

@ScottPJones speed. The GC tries very hard to be O(1) in the size of garbage memory (and it already fails at that in many regards). Adding a linear pass over all garbage is performance suicide, especially since typically a very small number of object needs finalization. If this is more of an UI question then we could have it look like this but we still would be storing an explicit list of live objects with finalizers somewhere.

**Comment by jakebolewski** [comment]
@StefanKarpinski I don't see how you set up that transitive relationship when object does not even exist yet.  Would you walk the object graph and then mark every reference that could reach a ref counted object every time one is constructed?

**Comment by tkelman** [comment]
Have wondered about this in the context of libxml bindings, where I think the most sane solution right now would be to implement manual refcounting. That hasn't been done yet so right now code that uses LightXML is either leaky or can have finalizers added that might be dangerously unaware of object relationships that are only directly observable through C.

**Comment by ScottPJones** [comment]
@carnaval I think that maybe there might be some sneaky means to not make it need a linear pass over all garbage...  I'll have to really dig into how the Julia GC works before I could say too much more.
I worry that the current method of having to have a list of objects with finalizers is also performance suicide...  it would seem that if the finalizer for a particular type of object were always `finalize!`, and the finalize! methods where written within the type declaration, like inner constructors, you might be able to segregate the allocation of objects that have finalization, so that would not be a problem... (or am I just being dense today? [could very well be!]

**Comment by carnaval** [comment]
@ScottPJones Sure we can do lots of things to make it faster than it is now. Segregation might be a pain since we use size-segregated pools so you will end up wasting quite a lot of memory if you only have a few serializable objects. Going through the list is not that much of a problem given that we actually have one of those lists per generation, so most of the time you are only checking the young one, which is mostly garbage, so you are mostly doing useful work. The costly thing here is probably that going through the list implies taking a (likely) cache miss for every object to check whether it is now garbage or not. If being finalizable was a property of the object's type we could have a hidden field per object storing its position in the finalizer list and keep a separate finalizable_alive bitfield so that, at the end of the collection, we could be lazy and only work on objects we are sure are now dead.

**Comment by carnaval** [comment]
I must add that much of the performance issue goes away if you use native finalizers instead of julia ones, since the julia ones requires keeping the object alive for one more collection.

**Comment by ScottPJones** [comment]
native finalizers?  Please explain...

**Comment by ScottPJones** [comment]
>If being finalizable was a property of the object's type we could have a hidden field per object storing its position in the finalizer list and keep a separate finalizable_alive bitfield so that, at the end of the collection, we could be lazy and only work on objects we are sure are now dead.

@carnaval That's exactly the sort of thing I was trying to get to... would that be hard to implement?
What are all the pros & cons?

**Comment by carnaval** [comment]
a quick google search revealed that it does not seem documented. oups.

The finalize function can accept a C function pointer instead of a julia function. The C function is then run in the middle of gc instead of at the end. Of course, if this C function touches the julia runtime, bad things can happen.

About the bitfield stuff, I don't see any obvious cons, except for the implementation effort of course. It should not be too hard to implement but be aware that the gc code is ... hem kind of messy. I would first start by putting up a benchmark were going through the fina list is actually the bottleneck (and not running the finalizers themselves).

If you start looking into this you may find lower hanging fruits however since I don't recall anyone spending too much time optimizing this.

**Comment by ScottPJones** [comment]
That will be easier then (using the native finalizers), of course now I have to go change my code again!  Thanks @carnaval!

>If you start looking into this you may find lower hanging fruits however since I don't recall anyone spending too much time optimizing this.

Well, I seem to already be doing well optimizing low hanging fruits in string handling, and that's another area where I've done similar stuff in the past, and might be able to help... (if I speed up parts of Julia enough, and [heaven forbid!] the wonderful startup I'm consulting for doesn't take off, maybe I can beg for a job at JC :grinning: [if the founders don't tar and feather me first!])  [if the startup does takes off, as I think it will, then next year I'm going to try to convince them to be a sponsor at JuliaCon]

**Comment by elextr** [comment]
@JeffBezanson you talked about using `finalize()` on BigInt, but who is managing the lifetime of that `BigInt`, ie who will call the `finalize()`.  For something like a `BigInt` I would have thought that the restriction to it only being used in a `with` is pretty untenable.

As @carnaval pointed out, reference counting can't be mixed with normal Julia references, reference counted objects must only be accessed via "counting references" and that would have to be enforced by the compiler.  The "counting references" then do the right thing when the *reference* is created, accessed, assigned, copied (parameter passing/return), and destructed (scope exit, including via exception).  IIUC currently not any/all those actions can be intercepted in Julia.

**Comment by JeffBezanson** [comment]
AFAICT BigInts cannot perform well if they depend on finalizers. The only way forward for BigInts is not to use finalizers at all.

**Comment by ScottPJones** [comment]
It sounds like BigInts (or at least portions of the code dealing with allocation/release) needs to be coded in Julia... wouldn't that remove the dependency on finalizers?

**Comment by JeffBezanson** [comment]
Yes, but we might not need to go that far. See #11202.

**Comment by davidanthoff** [comment]
It might be worth reading through the [.Net Dispose Pattern](https://msdn.microsoft.com/en-us/library/b1yfkh5e%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396). It essentially codifies much of what @JeffBezanson proposed, but takes a slightly different approach on banning finalizers altogether. The gist of it is that if an object has been disposed manually (i.e. its non managed resources have been freed by a call to ´´Dispose´´), then one can mark this object instance as not needing finalization anymore.

PS: This issue is strange to read, there are lots of responses to @ScottPJones, but not a single post from him?

**Comment by StefanKarpinski** [comment]
Not sure what happened but ScottPJones GitHub user account doesn't seem to exist anymore.

**Comment by carnaval** [comment]
So Scott's account was finalized ?

Sorry I'm out.

**Comment by tkelman** [comment]
Very strange, even PR's opened by him (like #11186 and others) are 404's now?

**Comment by mbauman** [comment]
Yikes.  I hope his employer didn't make him delete everything.  :(

I hope all is well.

**Comment by StefanKarpinski** [comment]
Seems that GitHub decided that Scott was a bot and removed all trace of him. He's contacted support and is trying to recover the account.

**Comment by MithrandirMiles** [comment]
I'm back, but just to respond to messages (I don't want to give up my ScottPJones account)

**Comment by MithrandirMiles** [comment]
I think GitHub has a very bad design, if when it makes somebody "invisible" (with no warning), it makes your entire history invisible, messages that other people might have wanted to see :-1: Where do we put issues on GitHub's design??? Grrrrr!
Thanks for all the very touching concern everybody!

**Comment by MithrandirMiles** [comment]
This is a copy of a now invisible message I sent a while ago (before realizing what GitHub had done to me!)

@davidanthoff I hope that wasn't a snide remark!  :grinning: I've been busy trying to 1) treat my wife to a nice Mother's Day, 2) get my PR #11186 as nice as possible, 3) get my "real" work with Julia done 4) figure out how to bribe any of the collaborators to merge in some of my Unicode fixes/performance improvements!
I've had a crazy idea... but I'm sure somebody here will shoot it down... besides having a finalize! method associated with a type, so that the box would just need to have a bit set to show that that object needs to be finalized by calling it's finalize! method, I was thinking you could also have maybe 16-bit reference counter, which starts at 1 when the object is allocated, it is only incremented if non-zero, so if it hits zero, it sticks there (meaning it will only get finalized when the GC gets to it), but if decremented to zero, the object is immediately finalized... (and the "need finalize" bit cleared)
I'm sure @JeffBezanson and @StefanKarpinski will tell me why I'm all wet... but that idea's been bouncing around my head all yesterday (could just be the drinks celebrating Mother's Day with my wife though!)

**Comment by davidanthoff** [comment]
@MithrandirMiles: no snide remark intended, I was just wondering what was going on. Glad this is "just" a github problem!

**Comment by MithrandirMiles** [comment]
@davidanthoff That's why I put the :grinning:!  (I've *tried* to increase my SNR around here... hopefully enough that you all will put up with me further!)

**Comment by JeffBezanson** [comment]
@MithrandirMiles this is terrible, I hope you get your account back!

The problems with refcounting are
- space to store the refcount
- the time spent updating it. all the loads and stores. especially in parallel.
- anything that requires iterating over dead objects to figure out which ones need to be finalized.

**Comment by MithrandirMiles** [comment]
@JeffBezanson I understand those points... that's why I was trying to think of crazy ideas, that could make @StefanKarpinski 's mixed ref count idea work... I hadn't had to worry so much about issues of parallelism with ref counts, we used a process model (up to 64K processes, all accessing shared memory for DB operations), but all of the compiler/interpreter's objects were just in each process... "objects", large strings, and top level variables were all ref counted...  About the iterating over the dead objects, I think various techniques *might* help, either segregating refcounted/finalizable objects in different pools, using bitmaps to see what pages had objects that needed to be looked at... but, like I said, I don't (yet!) know enough about the blood and guts of Julia to know what might be doable...

**Comment by wildart** [comment]
What about making object movable as in Rust? No referencing is allowed.

**Comment by vtjnash** [comment]
> Adding a linear pass over all garbage is performance suicide, especially since typically a very small number of object needs finalization

there are 2 unused bits in every pointer that the gc-allocates, due to the 16-byte alignment guarantee (assuming we get rid of the smallest pool). it seems like such a waste to not to be using them. but maybe @carnaval has something even better in mind for them then just marking which objects have finalizers?

**Comment by elextr** [comment]
> there are 2 unused bits in every pointer that the gc-allocates, due to the 16-byte alignment guarantee (assuming we get rid of the smallest pool). it seems like such a waste to not to be using them. but maybe @carnaval has something even better in mind for them then just marking which objects have finalizers?

Since using the lower bits for stuff means that each pointer load now needs a mask operation as well, it was found in another project I worked on to have a noticeable effect on performance.  Note, that project had a moving collector, so pointer re-loads may have been needed more often than they currently are in Julia.

But to impact a high frequency operation (pointer load) with a low frequency operation (like finalizers) might not be a net gain.

**Comment by timholy** [comment]
This topic just came up in https://github.com/JuliaLang/julia/pull/11280#issuecomment-104210821. If we use finalizers, it would be nice to use one of our precious tag bits to indicate whether an object's finalizer has already been run, so that all finalizer functions become safe to call twice.

**Comment by StefanKarpinski** [comment]
Another thing that might help here is the ability to add hooks for resource acquisition to trigger gc – e.g. being able to trigger a gc sweep when too many file handles are open.

**Comment by elextr** [comment]
Would need to force a full GC, which kind of defeats all the great effort put into the generational one.

**Comment by quinnj** [comment]
@JeffBezanson, so the usage of your proposal would be
```julia
type A
  # fields
end

close(a::A) = # closing stuff...

a = A()

# `with` block that guarantees finalization of object
with a
  # shenanigans
end

# which basically translates to
try
  # shenanigans with `a`
finally
    close(a)
end
```
What are the other differences with `with` vs. the `try-finally` block? Special-casing in GC? Would we still need machinery to flag whether a finalizer has been run or not (since each type will have a defined `close` method that could potentially be called manually)?

Thinking out loud, and having started reworking some `finalizer` work in SQLite recently, it would be nice to have the following pattern:

```julia
# create immutable Statement type where you can't tamper with the SQLite-provided handle
# also add the finalizer on creation, without an explicit `close` method to call (i.e. the only way to free is by finalization)
immutable SQLiteStatement
  handle::Ptr{Void}
  function SQLiteStatement(db::SQLiteDB)
    stmt = new(slite3_new_statement(db.handle))
    finalizer(stmt,()->sqlite3_finalize(stmt.handle))
    return stmt
  end
end

function do_stuff_with_statements()
  with SQLiteStatement(db) do stmt
    # do stuff with the `stmt`
  end
  # when the end of the `with` block is reached, the registered finalizer is run without having to do a full gc()
end
```
I think (perhaps wrongly) that's actually not too far from where we are now, but the key here is being able to ensure a finalizer is run once the object goes out of scope, with the necessity of not having to stop the world with a full gc().

**Comment by JeffBezanson** [comment]
Yes, the ideal situation seems to be freeing things eagerly using `with`, but with a safety net in the GC to ensure eventual finalization if somebody messes up. We have a `finalize` function now that does this. The only catch is performance:

* Adding the finalizers, and later having the GC go through them, is expensive.
* `finalize(x)` is currently O(n) in the number of registered finalizers. In any case it will be slower than a `close(x)` call (which could be inlined).

**Comment by quinnj** [comment]
How about this then for a somewhat more concrete proposal, (bringing a few different ideas together from this thread, #1037, [this comment](https://github.com/JuliaLang/julia/issues/11207#issuecomment-104240974), and #3067):

```julia
type A
  # fields
  # declare finalize method within type declaration
  finalize(a::A) = # what to do when finalized
end

# can manually create and finalize A()
a = A()
# do stuff with `a`
finalize(a) # doesn't "stop the world" for a full GC sweep, just calls finalize(`a`)

# use a `with` block for auto-finalization
# similar to `let-block` notation
with a = A(), b = A(), c = A()
  # do stuff with `a`, `b`, and `c`
end
# at end of block, finalize(a), finalize(b), and finalize(c) are called in order
```
There are a few things going on here:
* Declare the `finalize` method within the type declaration; this addresses #1037 by making `finalize` sticky to the type itself; this may also allow some optimizations when no `finalize` method is declared (since we'll know we'll never have to `finalize` that type)
* In place of the current `finalizer(x, func)`, `finalize(x)`, and custom `close` methods, we have a single, unified `finalize` method; a nice, simple interface
* Generally, this allows a separation of `finalize` from `gc()` since `finalize(x)` would avoid a full sweep
* Allows for flexibility in `finalize` usage (either by manually calling `finalize(x)`, or relying on a `with` block

From @timholy's [comment](https://github.com/JuliaLang/julia/issues/11207#issuecomment-104240974), we could also use a spare type bit to indicate if a finalizer has run on an object. This allows 1) avoiding calling `finalize` twice on potentially non-idempotent `finalize` methods and 2) *ensuring* `finalize` is called even if an object was created manually and not finalized manually.

**Comment by JeffBezanson** [comment]
To clarify, `finalize` does not do a full sweep, it's just not particularly efficient.

There are definitely some good points to that proposal. We could allocate finalizable objects in a separate arena, which would be annoying but more efficient than what we do now. Those object pools could have a bit array of finalize flags, to avoid using up a tag bit per object.

The only downside I see is that you can't do tricks like WeakKeyDict, which attaches finalizers to keys that remove them from the dictionary. That might be ok though, as that technique is not very efficient and should be replaced.

**Comment by StefanKarpinski** [comment]
When you say "tricks like WeakKeyDict", are there any other examples of using that trick that you're aware of? WeakKeyDict is a fairly special thing that it seems sensible to just support without needing to be able to *implement* it in the language.

**Comment by JeffBezanson** [comment]
The mmap code also attaches finalizers to Arrays to call munmap. That's about it I think.

**Comment by aviks** [comment]
In JavaCall, the object is created by the library, and returned to user code. The creator of the object does not control the lifecycle of the object... it necessarily escapes the creator scope. Upon creation, a finaliser is attached to the object that deallocates the JVM memory when the object is no longer required in Julia. 

If I understand @quinnj 's proposal to be that finaliser will no longer be called by the gc, then user code will be made responsible for managing the lifecycle of every object that it retrieves from the JavaCall boundary. That seems quite nasty to me.

**Comment by JeffBezanson** [comment]
No, it's just that you *could* efficiently bound the lifetime of something using `with`. If you don't use `with`, the GC will still clean up.

Is it sufficient for JavaCall to have finalizers associated with types, and not individual instances, e.g. a JavaObject type?

**Comment by ScottPJones** [comment]
:+1: to @quinnj 's proposal, it has what I had been asking for, the only thing is that I'd suggest calling the name `finalize!`, because it definitely modifies the object, and to avoid confusion with the old `finalize` method.

**Comment by aviks** [comment]
Ah, ok, thanks... I misunderstood. 

Yes, it should be sufficient to have finalisers associated with types. Currently, every object gets the same finaliser function. Of course, the type parameters and fields will need to be available to the finaliser.

**Comment by quinnj** [comment]
I wonder if the mmap and WeakKeyDict cases call for something like
```julia
finalize(a) do f
   # code to finalize `a` which is a type not declared with a `finalize` method
end
```
This wouldn't actually do the finalizing, just "move" `a` to the finalize pool of objects and the function argument would be run as the `finalize` method whenever that happens, either manually, from a `with` block, or when the object was destroyed.

Not sure how feasible "moving an object to the finalization pool of objects" would be though....

**Comment by amitmurthy** [comment]
Is https://github.com/JuliaLang/julia/issues/10960 then an artifact of the new gc? That could explain memory leaks with shared and distributed arrays. An ability to explicitly "free" remote objects will be quite useful, especially in cases where people are using distributed arrays across multiple hosts specifically to leverage every bit of memory available.

**Comment by ScottPJones** [comment]
@quinnj Carrying the discussion from #11280 over here, as requested...
You said:
>the problem with being able to call your own finalize! is you then need someway to tell if an object has been finalized or not.

That's precisely what I said I'd done, I have a pointer to something that needs to be finalized, so I simply set it to zero (C_NULL) in `finalize!`.  If you don't have a pointer, a flag can be used.  It is an extra check on each reference, but you stop having segfaults or problems with things outside of Julia being released.
It was the only way I could think of currently to make sure something can be finalized quickly most of the time, and still prevent memory (or other resource) leakage when things get GCed.
Do you have any better suggestions to handle that?

**Comment by amitmurthy** [comment]
Just noticed this when there are multiple finalizers defined for an object.

```
julia> type Foo
           v
       end

julia> f=Foo(0)
Foo(0)

julia> Foo(0)
Foo(0)

julia> for i in 1:10
           finalizer(f, x-> @schedule print("FINALIZED $i \n") )
       end

julia> f=nothing

julia> for i in 1:10
           print("calling gc for the $i th time\n");
           gc()
       end
calling gc for the 1 th time
calling gc for the 2 th time
FINALIZED 10 
FINALIZED 9 
calling gc for the 3 th time
FINALIZED 1 
FINALIZED 2 
calling gc for the 4 th time
FINALIZED 8 
calling gc for the 5 th time
FINALIZED 7 
calling gc for the 6 th time
FINALIZED 6 
calling gc for the 7 th time
FINALIZED 5 
calling gc for the 8 th time
FINALIZED 4 
calling gc for the 9 th time
FINALIZED 3 
calling gc for the 10 th time
```

Found it a little odd that all the finalizers are not executed together at the first `gc` itself.

**Comment by ScottPJones** [comment]
This wouldn't happen, if @timholy's idea (seconded by @quinnj [and myself]) to use a tag bit to say whether the finalizer had been run for an object.

**Comment by yuyichao** [comment]
@amitmurthy This is somewhat related to the (sub-)issue I noticed in https://github.com/JuliaLang/julia/issues/11814#issuecomment-114648637 . My guess is that running too many finalizers at the same time will cause a too long pulse but @carnaval should know for sure.

**Comment by yuyichao** [comment]
FWIW, the issue above in https://github.com/JuliaLang/julia/issues/11207#issuecomment-114620448 is solved by https://github.com/JuliaLang/julia/pull/13995 .

**Comment by amitmurthy** [comment]
Cool. And regarding the topic of this issue - It is not just about files, I don't think we have a choice but to use finalizers for remote references.  We can document that users can manually call `finalize` for better control on when remote resources get released, else it will only happen when `gc` eventually gets around to it.
